src/imputation_flags.py

```diff
@@ -0,0 +1,115 @@
+import numpy as np
+
+from src.flag_and_count_matched_pairs import flag_matched_pair_merge
+
+
+def create_impute_flags(df, target, reference, strata, auxiliary):
+    """
+    function to create logical columns for each type of imputation
+    output columns are needed to create the string flag column for
+    imputation methods.
+    Function requires f_predictive and b_predictive columns produced
+    by `flag_matched_pair` function
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        DataFrame containing forward, backward predictive period columns (
+        These columns are created by calling flag_matched_pair_merge forward
+        and backwards)
+
+    target : str
+        column name containing target variable
+    reference : str
+        column name containing business reference id
+    strata : str
+        column name containing strata information (sic)
+    auxiliary : str
+        column name containing auxiliary data
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe with four additional logical columns determining if target
+        is a return (r_flag) can be imputed by forward imputation (fir_flag),
+        backward imputation (bir_flag) or can be constructed (c_flag)
+    """
+    for direction in ["f", "b"]:
+        try:
+            df["{}_predictive_{}".format(direction, target)]
+        except KeyError:
+            raise KeyError(
+                "Dataframe needs column '{}_predictive_{}',".format(direction, target)
+                + " run flag_matched_pair function first"
+            )
+
+    df["f_predictive_" + target + "_roll"] = df.groupby([reference, strata])[
+        "f_predictive_" + target
+    ].ffill()
+    df["b_predictive_" + target + "_roll"] = df.groupby([reference, strata])[
+        "b_predictive_" + target
+    ].bfill()
+
+    df["r_flag"] = df[target].notna()
+
+    df["fir_flag"] = np.where(
+        df["f_predictive_" + target + "_roll"].notna() & df[target].isna(), True, False
+    )
+
+    df["bir_flag"] = np.where(
+        df["b_predictive_" + target + "_roll"].notna() & df[target].isna(), True, False
+    )
+
+    construction_conditions = df[target].isna() & df[auxiliary].notna()
+    df["c_flag"] = np.where(construction_conditions, True, False)
+
+    df = flag_matched_pair_merge(
```

Comment:
I think create_impute_flags should be a low level function, by calling flag_matched_pair_merge means that create_impute_flags is not a low funciton anymore. 

flag_matched_pair_merge would be called twice once outside of this function and once within, it might confuse the users, i would call it outside to call the f_predictive_auxiliary_roll and use it as a function parameter. Hope this made sense happy to chat in more detail in a call